### ISSUE TITLE
Why do we have z/Arch facilities enabled for ESA/390?!

### ISSUE THREAD

**Issue by Fish-Git** [Fish-Git/hyperion#74]
I've been working on [overhauling Hercules's facilities support](https://github.com/Fish-Git/hyperion/issues/71) and I noticed something that doesn't seem right to me.

In source file `feat390.h` there are more than a few z/Architecture facilities being enabled that, _technically_ (if I'm reading the [SA22-7201-08 (June 2003) ESA/390 Principles of Operation](https://www-05.ibm.com/e-business/linkweb/publications/servlet/pbi.wss?PAG=C11&SSN=17LHE0006164548724&TRL=TXT&WRD=&PBL=sa22-7201-08&LST=ALL&RPP=10&submit=Start+search) manual correctly), should _not_ be enabled.

Page 10-91 describes the behavior of the `STFL` _(Store Facility List)_ instruction and only documents facility bits _0-5_ and _16-20_.  It further states:

>   _"Bits 6-15 and 21-31 are reserved for indication
>    of new facilities.  The bits are currently stored
>    as zeros but may be stored as ones in the future."_

Yet, in header `feat390.h`, I'm seeing the following facilities being enabled:

```
   FEATURE_ETF2_ENHANCEMENT_FACILITY
   FEATURE_ETF3_ENHANCEMENT_FACILITY
   FEATURE_EXTENDED_TRANSLATION_FACILITY_3
   FEATURE_FPS_ENHANCEMENT
   FEATURE_FPS_EXTENSIONS
   INTERLOCKED_ACCESS_FACILITY_2
   FEATURE_MESSAGE_SECURITY_ASSIST_EXTENSION_3
   FEATURE_MESSAGE_SECURITY_ASSIST_EXTENSION_4
   FEATURE_STORE_FACILITY_LIST_EXTENDED_FACILITY
```

which correspond to facility list bits 24, 30, 22, 41, 37, 52, 76, 77 and 7, respectively.

_The enabling of the FEATURE_STORE_FACILITY_LIST_EXTENDED_FACILITY facility (bit 7) is especially puzzling (highly questionable IMHO)._

The [ESA/390 Principles of Operation](https://www-05.ibm.com/e-business/linkweb/publications/servlet/pbi.wss?PAG=C11&SSN=17LHE0006164548724&TRL=TXT&WRD=&PBL=sa22-7201-08&LST=ALL&RPP=10&submit=Start+search) doesn't even list the `STFLE` _(Store Facility List Extended)_ instruction as being a valid instruction for the architecture (and as demonstrated doesn't consider facility list bit 7 to be valid either).

**_Can someone please explain to me why these facilities are being enabled for ESA/390 when there's every indication that, according to available documentation, they *shouldn't* be?_**

Is there something I'm missing or not understanding correctly about Hercules that for some reason may require them to be enabled?

Would 390 guests suddenly stop working correctly if they were to be suddenly removed?

Inquiring minds would sure like to know! &nbsp; `:)`

Thanks.

**Comment by ivan-w** [comment]
Fish,

You may try removing the facility. But my guess is that *ANY* z/Architecture (older than something intended to be run on a z14) will no longer be able to IPL (since it IPLs in S/390 mode and then switches to z/Arch - and won't be able to issue a STFLE instruction when in S/390 mode prior to the z/Arch switch).
The following OSes may be affected : 
- z/VM
- z/OS
- z/VSE
- z/TPF
- z/Linux

**Comment by Fish-Git** [comment]
@ivan-w wrote:

> ... it IPLs in S/390 mode and then switches to z/Arch - and won't be able to issue a STFLE instruction when in S/390 mode prior to the z/Arch switch).

I would be _very_ surprised to see _any_ IBM operating system try executing an `STFLE` instruction _prior_ to switching to z/Architecture mode!

Don't ask me how I know but I happen to know for a fact that z/OS 2.3 switches to z/Architecture mode _first_ (almost _immediately_ after IPLing), and _then_, only once it's in z/Architecture mode, does it then execute ithe `STFLE` instruction and test for the existence of various facilities.

**Comment by Fish-Git** [comment]
This issue has already been fixed in my private Bitbucket repository and will eventually appear in my public GitHub repository just as soon as I'm finished with what I'm currently working on.

**Comment by Fish-Git** [comment]
Fixed by commit 4ee084f9e958ff5ae88c3ccad8be7500be47ebc5

**Comment by ivan-w** [comment]
May I enquire.. What's the purpose of bit 2 then ? (STFLE should be available on a system that implements z/Architecture but that is in ESA/390 architectural mode or in in ESA/390 compatibility mode)

**Comment by Fish-Git** [comment]
I don't understand your question/comment.

Facility bit 2 is defined as: "The z/Architecture architectural mode is active", and yes, STFL is an N3 instruction. What is your point?

**Comment by ivan-w** [comment]
The point is : STFLE should be available in ESA/390 mode if z/Architecture is installed (but not active). The point of Bit 2 is it can be 0 in this case (so unless it can never be 0, it makes no sense to actually have the bit).

**Comment by Fish-Git** [comment]
> STFLE should be available in ESA/390 mode if z/Architecture is installed

Why do you say that? STFLE is _not_ an N3 instruction, it is a _z/Architecture_ instruction, and thus should _not_ logically be available in ESA/390 mode (_regardless_ of whether z/Architecture happens to be installed or not).

You need to do a SIGP to switch to z/Architecture mode first (if STFL bit 1 is on) before you can execute the STFLE instruction. In ESA/390 mode, only STFL is available, and only provides information about ESA/390 facilities, but not about any z/Architecture facilities.

In order for the program to learn about z/Architecture facilities, it needs to first switch to z/Architecture (if STFL bit 1 is on) and then execute the z/Architecture STFLE instruction.

At least that's my understanding.

**Comment by Fish-Git** [comment]
> so unless it can never be 0, it makes no sense to actually have the bit

Sure it does. On the new z14, the bit is on as soon as you IPL. On other models it's not. This informs the program at IPL time whether or not it is already in z/Architecture mode, or whether it needs to manually switch to z/Architecture mode first (via SIGP).

What is so hard to understand about that?

**Comment by ivan-w** [comment]
What problem are you trying to solve by removing STFLE from the ESA/390 instruction set ?
It seems this is a solution in search of a problem !

**Comment by Fish-Git** [comment]
> What problem are you trying to solve by removing STFLE from the ESA/390 instruction set?

Proper behavior. Behavior that, based on my reading/understanding of the architecture, _should_ (I hope!) match that of real IBM hardware.

I don't have access to a real machine so I readly admit I don't know for sure whether the way I'm interpreting/implementing things is indeed the way a real machine behaves or not, but I'm guessing I am. If you have hard evidence/proof that my interpretation is incorrect, then by all means please provide such evidence/proof! Otherwise I see no point in arguing. It's obvious your opinion regarding this issue differs from mine and until we have proof one way or the other, who is to say who is right and who is wrong?

> It seems this is a solution in search of a problem!

Seems, yes. To you. But not to me. I see it as a _likely_ problem, so I fixed it. If you don't like (or agree with) what I've done, that's fine. You're entitle to your opinion just as I'm entitled to mine.

Fair enough?

**Comment by ivan-w** [comment]
Yes we disagree on in the interpretation of the PoP... However it DOES state : 

"A bit is set to one regardless of the current architectural
mode if its meaning is true. A meaning applies
to the current architectural mode unless it is said to
apply to a specific architectural mode."

and bit 7 says : 
7 The store-facility-list-extended facility is installed.

Note the VERY important fact that the sentence does NOT specify the familiar wording : "In the z/Architecture architectural mode."

Since STFL is a N3 instruction, it should work on a z/Architecture machine that is in ESA/390 (or ESA/390 compatibility mode). Bit 7 then should be set to 1 if the STFLE facility is installed - and since BIT 7 doesn't apply ONLY to z/Architecture mode, STFLE should also be available in ESA/390 mode (and bit 2 will be 0).

**Comment by ivan-w** [comment]
Ok... I finally managed to make a test on an actual IBM z13 (a 2964-NE1 running under z/VM 6.3 SL 1602)

- STFL returns this in ESA/390 mode : db6b fffb
- STFLE works and returns this in ESA/390 mode : db6b fffb fcff fd40 003c e000 0000 0000 c000
- STFL returns this after switching to z/Arch : fb6b fffb
- STFLE returns this after switching to z/Arch : fb6b fffb fcff fd40 003c e000 0000 0000 c000

--Ivan

**Comment by Fish-Git** [comment]
> As a hint: the z/VM 6.3 IPL issues STFL while still in ESA/390 mode - switches to z/Architecture mode and _**then**_ checks (from the result of the STFL instruction at +200 (0xC8) issued in ESA/390 mode) if bit 7 is 1 - and subsequently issues STFLE (in z/Architecture mode).

Well... yes, what you wrote is indeed true, but it's quite misleading.

Since I can't tell whether or not your intention was to be purposely misleading in order to advance your argument (claim) or whether it was just an honest (unintentional) mistake, I'm going to have to presume is was an honest mistake since I do afterall know you and am confident you would _never_ intentionally engage in such dishonesty. &nbsp; `;-)`

Isay this because I just completed my own test of a z/VM 6.3 IPL to try and confirm your findings, and my own results reach a different conclusion(*):

```
+++ STFL +++
+++ TM 0x0x000000c8, 0x40
+++ TM 0x0x000000c8, 0x20
+++ SIGP order 0x12 --> cpad 0x0000
+++ STFL +++
+++ TM 0x0x00000000000000c8, 0x01
+++ STFLE -->0x0x00000000004895a8
+++ MVC 0x0x0000000000492081( 64 ) <--- 0x0x00000000004895a8
+++ TM 0x0x00000000004895aa, 0x40
```

As you can see, it does indeed do a STFL to low core address 200 (0xC8) immediately after IPL while it's still in ESA/390 mode (before switching to z/Architecture mode), and _does_ indeed check for Facility bit 7 at low core address 200 (0xC8) (where the STFL instruction stores its Facility List) _after_ doing its SIGP to switch to z/Architecture mode, but what you failed to mention was, after it switches to z/Architecture it does yet _another_ STFL (while in _z/Architecture_ mode) and it is the results of _that_ STFL that it then checks for bit 7 being on or not, and only _then_ (after first verifying that STFLE is indeed supported based on the STFL done _**after**_ switching to z/Architecture mode) does it then issue its STFLE.

So, based on my own testing it appears z/VM 6.3 is coded to _not_ expect ESA/390 STFL bit 7 to be on, but _does_ expect _z/Arch_ STFL bit 7 to be on, which would seem to be evidence in _my_ favor. &nbsp; `:)`

So where does that leave us?

Answer: still in disagreement with one another unfortunately. &nbsp; `:(`

Man I wish we had acess to our own LPAR on a real z! We could write our own simple IPL routine and prove/settle this issue once and for all!

(And for the record I would be just as happy to be proven wrong as I would if proven right! My only interest is learning the _truth_, not with "winning ther argument"!)
______
`(*)` The above is the output of several Visual Studio tracepoints I set in each of the shown instructions so I'm rather confident in the accuracy of both the sequence abd the values displayed. I first inserted some debugging code into each of the instructions (to catch only what I was interested in) and then set the tracepoints on that code (i.e. to cause the debugger message to only be issued when the desired condition was true). The changes I had to make to Hercules was quite minor as can be seen in the attached patch:

* [Debug z/VM 6.3 STFL/SIGP/STFLE/TM sequence during IPL](https://github.com/Fish-Git/hyperion/files/1887527/fish_debug_stfle.zip)

**Comment by Fish-Git** [comment]
Here's the Hercules logfile output of the same thing:
```
2018/04/08 05:19:47.882 cmdtab.c(668)     HHC01603I ipl 0301
2018/04/08 05:19:47.892 cpu.c(1711)       HHC00811I Processor CP01: architecture mode ESA/390
2018/04/08 05:19:47.892 cpu.c(1711)       HHC00811I Processor CP03: architecture mode ESA/390
2018/04/08 05:19:47.893 cpu.c(1711)       HHC00811I Processor CP02: architecture mode ESA/390
2018/04/08 05:19:47.893 cpu.c(1711)       HHC00811I Processor IP04: architecture mode ESA/390
2018/04/08 05:19:47.893 cpu.c(1711)       HHC00811I Processor CP00: architecture mode ESA/390
2018/04/08 05:19:47.905 cmdtab.c(438)     HHC90000D DBG: RC = 0
2018/04/08 05:19:47.991 cckddasd.c(1436)  HHC00107I Starting thread cckd_ra(), active=0, started=0, max=2
2018/04/08 05:19:47.993 cckddasd.c(1436)  HHC00107I Starting thread cckd_ra(), active=0, started=1, max=2
2018/04/08 05:19:47.996 cckddasd.c(1518)  HHC00100I Thread id 00001b9c, prio  0, name Read-ahead thread-1 started
2018/04/08 05:19:47.997 cckddasd.c(1518)  HHC00100I Thread id 00000834, prio  0, name Read-ahead thread-2 started
2018/04/08 05:19:48.353 esame.c(5527)     +++ STFL +++
2018/04/08 05:19:49.404 general2.c(1502)  +++ TM +++  0xC8, X'40'
2018/04/08 05:19:50.511 general2.c(1502)  +++ TM +++  0xC8, X'20'
2018/04/08 05:19:51.511 control.c(5546)   +++ SIGP +++
2018/04/08 05:19:52.565 control.c(6226)   HHC00814I Processor CP00: SIGP Set architecture mode            (12) CP00, PARM 00000001: CC 0
2018/04/08 05:19:52.566 cpu.c(1711)       HHC00811I Processor CP00: architecture mode z/Arch
2018/04/08 05:19:52.566 esame.c(5527)     +++ STFL +++
2018/04/08 05:19:53.671 general2.c(1502)  +++ TM +++  0xC8, X'01'
2018/04/08 05:19:54.724 esame.c(5586)     +++ STFLE +++  -->  0x4895A8
2018/04/08 05:19:55.779 general1.c(5189)  +++ MVC +++  0x492081( 64 ) <-- 0x4895A8
2018/04/08 05:19:56.834 general2.c(1502)  +++ TM +++  0x4895AA, X'40'
```

As you can plainly see, the STFL is executed **_again_,_after_** switching to z/Arch mode, and it is the resultsof _that_ STFL that it's using to check if Facility bit 7 is on or not (so it can know whether or not it can safely issue the STFLE instruction or not).

Like I said: evidence in _my_ favor. Inconclusive yes, but evidence nonetheless.

**Comment by ivan-w** [comment]
Fish,

You can not have *low level* access to a z13 because it always runs in LPAR mode (basic mode is not available on anything newer than a z800). so it always runs PR/SM which also uses SIE.

Again when STFL is executed in ESA/390 mode, bit 7 is 1 so it DOES mean : STFLE is available in the *current architecture mode* (ESA/390)

Furthermore, I would  be very enticed to trust z/VM to provide an architecturally correct view of the system.

And I checked.. z/VM 6.3 doesn't intercept the STFL instruction, so it is executed on the next lower level (which in the case of my test case is PR/SM - which I trust even more than z/VM to be architecturally correct). So when you execute STFL under z/VM, it is actually executed by the underlying hardware (not simulated by z/VM).

And I still claim : if the intend of the Principle of Operations was to indicate that STFLE was only available in the z/Architecture mode, bit 7 *WOULD* state : 'In the z/Architecture mode" (which is NOT the case).

My evidences trump yours (not THAT trump mind it)...

--Ivan

**Comment by Fish-Git** [comment]
Ivan,

As long as you're 100% certain that z/VM 6.3 doesn't intercept the STFL instruction (how were you able to determine that by the way?) then fine, I'm convinced.

I believe the test you did is probably the closest we can get to the hardware given the current circumstances. I apologize for being so obstinate, but I wanted to be sure! _Thank you_ for your extraordinarily patience.

I'll be committing the appropriate changes shortly.

Interestingly, after reviewing the bit setting in light of the all-important key phrase that you kindly reminded me about (_thank you again for that!_):

> "A meaning applies to the current architectural mode
> unless it is said to apply to a specific architectural mode."

I found several other bits that were wrong as well (bit 52, interestingly, being one of them).

I'm glad to _finally_ put this issue to rest once and for all, and again apologize most sincerely for my being so obstinate for so long.

But we're cool now. &nbsp; `:)`

Thanks again.

**Comment by Fish-Git** [comment]
(Re-opening until I commit the final fix, at which time I'll close it again.)

**Comment by ivan-w** [comment]
Fish,
We've always been cool ! No worries there.

Now about STFL not being intercepted : I looked at how z/VM 6.3 behaves under hercules and it doesn't intercept STFL when it is executed in a virtual machine (therefore running under SIE).

The only reason hercules would have to intercept STFL and/or STFLE would be to clear some bits in the facility list in case z/VM is operating in SSI mode and the virtual machine is within a migration domain in which not all members have the same capabilities. This is the reason of the term of the PoP : 

```
STORE FACILITY LIST and STORE FACILITY LIST
EXTENDED may report the absence of a facility,
even though the execution of the facility’s instructions
may appear to indicate that the facility is installed.
The under-reporting of facility indications may occur
because one or more systems in the computing environment
(to which the application may be relocated)
do not have the facility installed.
```
Anyway, great work dude... I've started looking more closely at the SDL version, and it's looking good !

--Ivan

**Comment by Fish-Git** [comment]
> We've always been cool ! No worries there.

I figured as much but I wanted to be sure. As I mentioned I **_do_** realize how stubborn/obstinate I have been regarding this issue and I wasn't sure whether I was pissing you off or not. Glad to hear I wasn't (didn't).

> The only reason _**hercules** (sic)_ would have to intercept STFL and/or STFLE would be to clear some bits in the facility list in case z/VM is operating in SSI mode and the virtual machine is within a migration domain in which not all members have the same capabilities.

I believe you mean z/VM, not Hercules? (Yes there's code in Hercules to deal with this, but the SIE state block that tells Hercules to do it comes from z/VM, yes?)

Anyway, yes, I too _also_ saw in the PoP what you mentioned:

> STORE FACILITY LIST and STORE FACILITY LIST
> EXTENDED may report the absence of a facility,
> even though the execution of the facility’s instructions
> may appear to indicate that the facility is installed.
> The under-reporting of facility indications may occur
> because one or more systems in the computing environment
> (to which the application may be relocated)
> do not have the facility installed.


In fact, the Hecules code that handles it in the `start_interpretive_execution` SIE instruction function was recently tweaked by me (where the `fld` Facility List Designator is processed that defines a mask for disabling SIE guest facilities which are normally enabled on the SIE host). It was one of the things I was looking at relating to this whole STFL/STFLE issue that I originally missed. So I did another test to be sure. It was only after I did that test that was finally ready to concede your point: STFLE (and several other facilities too) **_are_** indeed available in ESA/390 mode too, and thus should have their bits set.

> Anyway, great work dude... I've started looking more closely at the SDL version, and it's looking good !

Thanks! It was quite a challenge getting it done. It keep snowballing, one thing leading to another, and I was beginning to wonder if I was _ever_ going to finish! &nbsp; `:)`

But I think everything is straightened out now. If you see anything suspicious or questionable, please _**do**_ let me know! I greatly value and very much appreciate your high-quality and thoughful feedback.

Thanks again, Ivan!

--Fish

**Comment by Fish-Git** [comment]
Closed (for good this time!) by commit 5ae064486f607ac11aa7ec8a32e60aa2902d5fc2
